### PR TITLE
Rename conflicting identifiers

### LINKED ISSUE #628: Avoid name collisions on C code
- [ ] Avoid two functions with the same name -> _remane_. If both are `<api>` then throw an error.
- [ ] Same name in local variable names and functions. We can rename the local variables using `_` prefix.

### PR DESCRIPTION
**Pull request by ivojawer**
closes #628 

Checks and renames this conflicts:

- Selector vs other selectors when translated to functions
- Locals vs other selectors when translated to functions

### CODE REVIEW

File: smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st [hunk 1]
```diff
@@ -39,6 +39,17 @@ SlangBasicTranslationTest >> defineAtCompileTime: aString [
...
+{ #category : #helpers }
+SlangBasicTranslationTest >> functionIdentifier: aMethodTranslation [
+	| prototype matchIdentifier |
+	
+	prototype := aMethodTranslation lineNumber: 3.
+	
+	matchIdentifier := '([a-zA-Z0-9_])+\(' asRegex.
+	
+	^((matchIdentifier matchesIn: prototype) first) copyWithout: $(
+]
```
Comment (PalumboN) [inline smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st:51]: Uufff

File: smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st [hunk 2]
```diff
@@ -1779,6 +1800,23 @@ methodWithoutReturn(void)
...
+	tMethod1 := self getTMethodFrom: #methodWithRepeatedCFunctionName.
+	tMethod2 := generator methods at: #methodWithRepeatedCFunctionName:.
+
+	generator renameConflictingMethods.
+
+	translation1 := self translate: tMethod1.
+	translation2 := self translate: tMethod2.
+	
+	self assert: (self functionIdentifier: translation1)
+		~= (self functionIdentifier: translation2)
```
Comment (PalumboN) [inline smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st]:
```suggestion
	self 	deny: (self functionIdentifier: translation1)
		equals: (self functionIdentifier: translation2)
```

File: smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st [hunk 3]
```diff
@@ -6557,6 +6595,23 @@ SlangBasicTranslationTest >> testVariableUpdateAssignmentPlus [
...
+	method := self getTMethodFrom: #methodDefiningConflictingVariable.
+
+	generator renameConflictingLocals.
+
+	translation := self translate: method.
+
+
+	self
+		assert: (translation lineNumber: 5)
+		equals: '	char *l_methodWithoutReturn;'
```
Comment (PalumboN) [inline smalltalksrc/Slang-Tests/SlangBasicTranslationTest.class.st]:
```suggestion
	self assert: (translation includesSubstring: 'char *l_methodWithoutReturn').
```
?

File: smalltalksrc/Slang/CCodeGenerator.class.st [hunk 4]
```diff
@@ -4415,7 +4416,61 @@ CCodeGenerator >> removeVariable: aName ifAbsent: ifAbsentBlock [
 	^variables remove: aName ifAbsent: ifAbsentBlock
 ]
 
-{ #category : #translating }
+{ #category : #renaming }
+CCodeGenerator >> renameConflictingLocals [
```
Comment (PalumboN) [inline smalltalksrc/Slang/CCodeGenerator.class.st:4423]: Call to this method is missing

File: smalltalksrc/Slang/CCodeGenerator.class.st [hunk 5]
```diff
@@ -4415,7 +4416,61 @@ CCodeGenerator >> removeVariable: aName ifAbsent: ifAbsentBlock [
...
+		cFunctionNames at: functionName put: (conflicts copyWith: m) ].
+
+	toRename := cFunctionNames select: [ :conflicts | conflicts size > 1 ].
+
+	toRename keysDo: [ :key |
+		| conflictingMethods |
+		conflictingMethods := toRename at: key.
+
+		"api methods cannot be renamed"
+		(conflictingMethods select: [ :m | m isAPIMethod ]) size > 1
```
Comment (PalumboN) [inline smalltalksrc/Slang/CCodeGenerator.class.st:4457]:
```suggestion
		(conflictingMethods anySatisfy: [ :m | m isAPIMethod ])
```

File: smalltalksrc/Slang/CCodeGenerator.class.st [hunk 6]
```diff
@@ -4415,7 +4416,61 @@ CCodeGenerator >> removeVariable: aName ifAbsent: ifAbsentBlock [
...
+			ifTrue: [
+				TranslationError signal:
+					'Conflicting API methods would be translated to "' , key , '"' ].
+
+
+		(conflictingMethods reject: [ :m | m isAPIMethod ]) do: [ :method |
+			method isStructAccessor ifFalse: [
+				self
+					addSelectorTranslation: method selector
+					to: key , method args size asString ] ] ]
```
Comment (PalumboN) [inline smalltalksrc/Slang/CCodeGenerator.class.st]: Some comments:

1. The `reject:` is not needed, if exists this method throws an exception before.
2. If struct accessors do not care for the collision, maybe could be filtered at the beginning
3. This translation does work if the collision is between methods with the same number of args (aka polymorphism coming soon!). This is not a problem, we just need to open another issue.
4. We can rewrite this method in this way (it's just a suggestion)
```st
cFunctionNames := IdentityDictionary new.
methods
	reject: [:m | m isStructAccessor] 
	thenDo: [:m | ... fil the dictionary ... ].
toRename := cFunctionNames select: [ :conflicts | conflicts size > 1 ]. 
toRename keysAndValuesDo: [ :key :conflictingMethods | 
	conflictingMethods do: [:method | 
		method isAPIMethod ifTrue: [ ...signal ... ]
		self
			addSelectorTranslation: method selector
			to: key , method args size asString
	]
]
```

File: smalltalksrc/Slang/CCodeGenerator.class.st [hunk 7]
```diff
@@ -4432,7 +4487,7 @@ CCodeGenerator >> renameKeywordsConflicts [
 			method renameVariablesUsing: localRewording ].
 
 		"translate selector"
-		cSelector := self cSelectorName: method selector.
+		cSelector := self cFunctionNameFor: method selector.
```
Comment (PalumboN) [inline smalltalksrc/Slang/CCodeGenerator.class.st:4487]: 🙌

**Review by PalumboN — APPROVED** [review]
Niceeeeee!

I left some comments 🔍 👁️

File: smalltalksrc/Slang/CCodeGenerator.class.st (hunk 4 above)
Comment (ivojawer) [reply smalltalksrc/Slang/CCodeGenerator.class.st:4423]: yep, added the call in `emitCCode`

**Review by guillep — APPROVED** [review]
Very good job @ivojawer thanks!!!!

**Pull request merged**